src/Common/Scheduler/ResourceGuard.h

```diff
@@ -63,43 +113,59 @@ class ResourceGuard
 
         void wait()
         {
+            CurrentMetrics::Increment scheduled(metrics->scheduled_count);
+            auto timer = CurrentThread::getProfileEvents().timer(metrics->wait_microseconds);
             std::unique_lock lock(mutex);
             dequeued_cv.wait(lock, [this] { return state == Dequeued; });
         }
 
-        void finish()
+        void finish(ResourceCost real_cost_, ResourceLink link_)
         {
             // lock(mutex) is not required because `Dequeued` request cannot be used by the scheduler thread
             chassert(state == Dequeued);
             state = Finished;
+            if (estimated_cost != real_cost_)
+                link_.queue->adjustBudget(estimated_cost, real_cost_);
```

Comment:
There was a bug. All ResourceGuard were used in the wrong way passing `request->cost` into `adjustBudget()` but `enqueueRequestUsingBudget` changes `request->cost` before enqueing the request, so the correct initial estimated cost was not stored anywhere, but it is required for correct adjustment of the budget. Now we store it in ResourceGuard and it's usage is much simpler and less error-prone.